### PR TITLE
Issue Transaction Test Suite

### PR DESCRIPTION
**Pull request by irina-pereiaslavskaia**
New positive and negative tests for Issue Tx.

### CODE REVIEW

File: itests/issue_transaction_test.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+		tx, errGo, errScala := issue(suite, td, timeout)
+		txIds[name] = tx.ID
+
+		currentBalanceInWaves := utl.GetAvalibleBalanceInWaves(&suite.BaseSuite, td.Account.Address)
+		actualBalanceInWaves := initBalanceInWaves - currentBalanceInWaves
+		actualAssetBalance := utl.GetAssetBalance(&suite.BaseSuite, td.Account.Address, tx.ID.Bytes())
+
+		expectedBalanceInWaves, _ := strconv.ParseInt(td.Expected["waves diff balance"], 10, 64)
+		expectedAssetBalance, _ := strconv.ParseInt(td.Expected["asset balance"], 10, 64)
```
Comment (nickeskov) [inline itests/issue_transaction_test.go:112]: The same as above.

File: itests/testdata/issue_tx_testdata.go [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+	b := make([]byte, n)
+	for j := range b {
+		b[j] = letterBytes[rand.Intn(len(letterBytes))]
+	}
+	return string(b)
+}
+
+func GetCurrentTimestampInMs() uint64 {
+	return uint64(time.Now().UnixNano() / 1000000)
+}
```
Comment (nickeskov) [inline itests/testdata/issue_tx_testdata.go:56]: You can use `time.Now().UnixMilli()`.
```suggestion
func GetCurrentTimestampInMs() uint64 {
	return uint64(time.Now().UnixMilli())
}
```

File: itests/issue_transaction_test.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+type IssueTxSuite struct {
+	fixtures.BaseSuite
+}
+
+func newSignIssueTransaction(suite *IssueTxSuite, testdata testdata.IssueTestData) *proto.IssueWithSig {
+	tx := proto.NewUnsignedIssueWithSig(testdata.Account.PublicKey, testdata.AssetName,
+		testdata.AssetDesc, testdata.Quantity, testdata.Decimals, testdata.Reissuable, testdata.Timestamp, testdata.Fee)
+	err := tx.Sign(testdata.ChainID, testdata.Account.SecretKey)
+	suite.NoError(err, "failed to create proofs from signature")
```
Comment (nickeskov) [inline itests/issue_transaction_test.go:24]: Use `suite.Require()` for fatal errors.

**Review by nickeskov — CHANGES_REQUESTED** [review]

**Review by alexeykiselev — CHANGES_REQUESTED** [review]

File: itests/issue_transaction_test.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
+package integration
+
+import (
+	"github.com/stretchr/testify/suite"
+	"github.com/wavesplatform/gowaves/itests/fixtures"
+	"github.com/wavesplatform/gowaves/itests/net"
+	"github.com/wavesplatform/gowaves/itests/testdata"
+	utl "github.com/wavesplatform/gowaves/itests/utilities"
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/issue_transaction_test.go:13]: Package was renamed

File: itests/utilities/issue_tx_utility.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	return suite.Clients.GoClients.GrpcClient.GetAssetBalance(suite.T(), address, id).GetAmount()
+}
+
+func GetHeightGo(suite *integration.BaseSuite) *client.BlocksHeight {
+	return suite.Clients.GoClients.GrpcClient.GetHeight(suite.T())
+}
+
+func GetHeightScala(suite *integration.BaseSuite) *client.BlocksHeight {
+	return suite.Clients.ScalaClients.GrpcClient.GetHeight(suite.T())
+}
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/utilities/issue_tx_utility.go]: This functions were removed

File: itests/utilities/issue_tx_utility.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/wavesplatform/gowaves/pkg/crypto"
+	"github.com/wavesplatform/gowaves/pkg/proto"
+	"time"
+)
+
+func GetAccount(suite *integration.BaseSuite, i int) config.AccountInfo {
+	return suite.Cfg.Accounts[i]
+}
+
+func GetAvalibleBalanceInWaves(suite *integration.BaseSuite, address proto.WavesAddress) int64 {
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/utilities/issue_tx_utility.go]: Done

File: itests/utilities/issue_tx_utility.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+		_, _, errGo := suite.Clients.GoClients.HttpClient.TransactionInfoRaw(*id)
+		_, _, errScala := suite.Clients.ScalaClients.HttpClient.TransactionInfoRaw(*id)
+		if errGo == nil {
+			txIds["Go "+name] = id.String()
+		}
+		if errScala == nil {
+			txIds["Scala "+name] = id.String()
+		}
+	}
+	return txIds
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/utilities/issue_tx_utility.go]: Renamed

File: itests/node_client/node_client.go [hunk 8]
```diff
@@ -85,15 +85,14 @@ func retry(timeout time.Duration, f func() error) error {
 	return nil
 }
 
-func (c *NodesClients) WaitForTransaction(t *testing.T, ID *crypto.Digest, timeout time.Duration) {
-	err := retry(timeout, func() error {
+func (c *NodesClients) WaitForTransaction(t *testing.T, ID *crypto.Digest, timeout time.Duration) (error, error) {
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/node_client/node_client.go]: Fix

File: itests/issue_transaction_test.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+		tx1, errGo1, errScala1 := issue(suite, td, timeout)
+		tx2, errGo2, errScala2 := issue(suite, testdata.DataChangedTimestamp(&td), timeout)
+
+		currentBalanceInWaves := utl.GetAvalibleBalanceInWaves(&suite.BaseSuite, td.Account.Address)
+		actualDiffBalanceInWaves := initBalanceInWaves - currentBalanceInWaves
+		actualAsset1Balance := utl.GetAssetBalance(&suite.BaseSuite, td.Account.Address, tx1.ID.Bytes())
+		actualAsset2Balance := utl.GetAssetBalance(&suite.BaseSuite, td.Account.Address, tx2.ID.Bytes())
+		diffBalanceInWaves, _ := strconv.ParseInt(td.Expected["waves diff balance"], 10, 64)
+		expectedDiffBalanceInWaves := 2 * diffBalanceInWaves
+		expectedAssetBalance, _ := strconv.ParseInt(td.Expected["asset balance"], 10, 64)
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/issue_transaction_test.go]: Done

File: itests/issue_transaction_test.go (hunk 1 above)
Comment (irina-pereiaslavskaia) [reply itests/issue_transaction_test.go]: Done

File: itests/testdata/issue_tx_testdata.go (hunk 2 above)
Comment (irina-pereiaslavskaia) [reply itests/testdata/issue_tx_testdata.go]: Done

File: itests/issue_transaction_test.go (hunk 3 above)
Comment (irina-pereiaslavskaia) [reply itests/issue_transaction_test.go]: Done

File: itests/integration_test.go [hunk 10]
```diff
@@ -1,99 +1,51 @@
-package integration_test
+package integration
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/integration_test.go]: Fix

File: itests/integration_test.go [hunk 11]
```diff
@@ -1,99 +1,51 @@
...
 
 import (
-	"context"
-	"strings"
+	"github.com/wavesplatform/gowaves/itests/fixtures"
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/integration_test.go]: Done

File: itests/integration_test.go [hunk 12]
```diff
@@ -1,99 +1,51 @@
...
-}
-
-func (suite *ItestSuite) TearDownTest() {
-	suite.clients.SendEndMessage(suite.T())
+type SendTxSuite struct {
+	fixtures.BaseSuite
 }
 
-func (suite *ItestSuite) Test_SendTransaction() {
+func (suite *SendTxSuite) Test_SendTransaction() {
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/integration_test.go]: File was removed.

File: itests/integration_test.go [hunk 13]
```diff
@@ -1,99 +1,51 @@
...
 func TestItest1Suite(t *testing.T) {
 	t.Parallel()
-	suite.Run(t, new(ItestSuite))
+	suite.Run(t, new(SendTxSuite))
 }
 
 func TestItest2Suite(t *testing.T) {
 	t.Parallel()
-	suite.Run(t, new(ItestSuite))
+	suite.Run(t, new(SendTxSuite))
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/integration_test.go]: File was removed.

File: itests/utilities/issue_tx_utility.go [hunk 14]
```diff
@@ -0,0 +1,46 @@
+package utl
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/utilities/issue_tx_utility.go]: Done

File: itests/utilities/issue_tx_utility.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+func GetAvalibleBalanceInWavesGo(suite *i.BaseSuite, address proto.WavesAddress) int64 {
+	return suite.Clients.GoClients.GrpcClient.GetWavesBalance(suite.T(), address).GetAvailable()
+}
+
+func GetAssetBalanceGo(suite *i.BaseSuite, address proto.WavesAddress, id []byte) int64 {
+	return suite.Clients.GoClients.GrpcClient.GetAssetBalance(suite.T(), address, id).GetAmount()
+}
+
+func GetTxIdsInBlockchain(suite *i.BaseSuite, ids map[string]*crypto.Digest, timeout time.Duration) map[string]string {
+	time.Sleep(timeout)
```
Comment (nickeskov) [inline itests/utilities/issue_tx_utility.go:26]: Btw, it's ok to have `Sleep` here? Looks like unnecessary slowdown of test. You can wait and range through all in parallel.

File: itests/issue_transaction_test.go [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+import (
+	"github.com/stretchr/testify/suite"
+	"github.com/wavesplatform/gowaves/itests/fixtures"
+	"github.com/wavesplatform/gowaves/itests/net"
+	"github.com/wavesplatform/gowaves/itests/testdata"
+	utl "github.com/wavesplatform/gowaves/itests/utilities"
+	"github.com/wavesplatform/gowaves/pkg/crypto"
+	"github.com/wavesplatform/gowaves/pkg/proto"
+	"strconv"
```
[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/issue_transaction_test.go]: Done

File: itests/issue_transaction_test.go [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+	bts, err := tx.MarshalBinary()
+	suite.NoError(err, "failed to marshal tx")
+	txMsg := proto.TransactionMessage{Transaction: bts}
+
+	suite.Conns = net.Reconnect(suite.T(), suite.Conns, suite.Ports)
+	suite.Conns.SendToEachNode(suite.T(), &txMsg)
+
+	errGo, errScala := suite.Clients.WaitForTransaction(suite.T(), tx.ID, timeout)
+	return errGo, errScala
+}
```
[the remark this answers is not in the document]
Comment (nickeskov) [reply itests/issue_transaction_test.go]: Still need to be fixed

[the remark this answers is not in the document]
Comment (irina-pereiaslavskaia) [reply itests/issue_transaction_test.go]: Done

**Pull request merged**